ath79: add support for ASUS PL-AC56

therealsummoner opened the pull request:
Asus PL-AC56 Powerline Range Extender Rev.A1
(in kit with Asus PL-E56P Powerline-slave)

Hardware specifications:
Board: AP152
SoC: QCA9563 2.4G n 3x3
PLC: QCA7500
WiFi: QCA9882 5G ac 2x2
Switch: QCA8337 3x1000M
Flash: 16MB 25L12835F SPI-NOR
DRAM SoC: 64MB w9751g6kb-25
DRAM PLC: 128MB w631gg6kb-15

Clocks: CPU:775.000MHz, DDR:650.000MHz, AHB:258.333MHz, Ref:25.000MHz

MAC addresses as verified by OEM firmware:
use            address      source
Lan/Wan/PLC    *:10         art 0x1002 (label)
2G             *:10         art 0x1000
5G             *:14         art 0x5000

Important notes:
 - the PLC firmware has to be provided and copied manually onto the
    device! The PLC here has no dedicated flash, thus the firmware file
    has to be uploaded to the PLC controller at every system start
 - the PLC functionality is managed by the script /etc/init.d/plc_basic,
    a very basic script based on the the one from Netadair (netadair dot de)

Installation:

Asus windows recovery tool:
 - have to have the latest Asus firmware flashed before continuing!
 - install the Asus firmware restoration utility
 - unplug the router, hold the reset button while powering it on
 - release when the power LED flashes slowly
 - specify a static IP on your computer:
    IP address: 192.168.1.75
    Subnet mask 255.255.255.0
 - start the Asus firmware restoration utility, specify the factory image
    and press upload
 - do NOT power off the device after OpenWrt has booted until the LED flashing

TFTP Recovery method:
 - have to have the latest Asus firmware flashed before continuing!
 - set computer to a static ip, 192.168.1.75
 - connect computer to the LAN 1 port of the router
 - hold the reset button while powering on the router for a few seconds
 - send firmware image using a tftp client; i.e from linux:
    $ tftp
    tftp> binary
    tftp> connect 192.168.1.1
    tftp> put factory.bin
    tftp> quit
 - do NOT power off the device after OpenWrt has booted until the LED flashing

Additional notes:
 - the pairing buttons have to have pressed for at least half a second,
    it doesn't matter on which plc device (master or slave) first
 - it is possible to pair the devices without the button-pairing requirement
    simply by pressing reset on the slave device. This will default to the
    firmware settings, which is also how the plc_basic script is setting up
    the master device, i.e. configuring it to firmware defaults
 - the PL-E56P slave PLC has its dedicated 4MByte SPI, thus it is capable
    to store all firmware currently available. Note that some other
    slave devices are not guarantied to have the capacity for the newer
    ~1MByte firmware blobs!
    To have a good overlook about the slave device, here are its specs:
    same QCA7500 PLC controller, same w631gg6kb-15 128MB RAM,
    25L3233F 4MB SPI-NOR and an AR8035-A 1000M-Transceiver

Signed-off-by: Tamas Balogh <tamasbalogh@hotmail.com>

hauke commented on target/linux/ath79/image/generic.mk:
```diff
@@ -375,6 +375,21 @@ define Device/aruba_ap-105
+	append-rootfs | pad-rootfs
+  DEVICE_PACKAGES := kmod-ath10k-ct-smallbuffers ath10k-firmware-qca988x-ct \
+	open-plc-utils open-plc-utils-hpavkey open-plc-utils-modpib \
+	open-plc-utils-plchost open-plc-utils-plctool
```
The 'open-plc' tools are only available in the package feed, I think we should not add them by default to the images.

therealsummoner replied:
Can be better alright, thanx, I'll take them out.

hauke commented:
Please rebase and remove the merge commits.

therealsummoner commented:
Finished preparing the plc functionality how-to on the device's temporary page:
https://openwrt.org/inbox/toh/asus/pl-ac56_a1#notes
I'll transfer this after it gets supported to the live device page.

hauke commented on target/linux/ath79/dts/qca9563_asus_pl-ac56.dts:
```diff
@@ -0,0 +1,156 @@
+		qca,mib-poll-interval = <500>;
+
+		qca,ar8327-initvals = <
+			0x04 0x80000080 /* AR8327_REG_PAD0_MODE */
```
What is this bit doing?
On the AR8327 it is a reserved bit.
https://raw.githubusercontent.com/Deoptim/atheros/master/AR8327.pdf

therealsummoner replied:
This is the result of the register dump from the OEM FW, made like this in ar8327.c :

`static int
ar8327_hw_config_of(struct ar8xxx_priv *priv, struct device_node *np)
{
	struct ar8327_data *data = priv->chip_data;
	const __be32 *paddr;
	int len;
	int i;

pr_info("0x04 %08x\n", ar8xxx_read(priv, AR8327_REG_PAD0_MODE));
pr_info("0x08 %08x\n", ar8xxx_read(priv, AR8327_REG_PAD5_MODE));
pr_info("0x0c %08x\n", ar8xxx_read(priv, AR8327_REG_PAD6_MODE));
pr_info("0x10 %08x\n", ar8xxx_read(priv, AR8327_REG_POWER_ON_STRAP));
pr_info("0x50 %08x\n", ar8xxx_read(priv, AR8327_REG_LED_CTRL0));
pr_info("0x54 %08x\n", ar8xxx_read(priv, AR8327_REG_LED_CTRL1));
pr_info("0x58 %08x\n", ar8xxx_read(priv, AR8327_REG_LED_CTRL2));
pr_info("0x5c %08x\n", ar8xxx_read(priv, AR8327_REG_LED_CTRL3));
pr_info("0x7c %08x\n", ar8xxx_read(priv, AR8327_REG_PORT_STATUS(0)));

	paddr = of_get_property(np, "qca,ar8327-initvals", &len);
	if (!paddr || len < (2 * sizeof(*paddr)))
		return -EINVAL;
.
.
.`
I'd say Asus had here either more/newer info about the registers, or they set up the bit for e.g. internal use, but this I cannot tell. I was however focusing to represent the registers' contents as original as possible.

hauke replied:
ok thanks fir the explanation.

therealsummoner commented:
Does this look ok for pulling yet?
Functionally its all ok, I'm using the device since months.

therealsummoner committed:
ath79: add support for ASUS PL-AC56

Asus PL-AC56 Powerline Range Extender Rev.A1
(in kit with Asus PL-E56P Powerline-slave)

Hardware specifications:
Board: AP152
SoC: QCA9563 2.4G n 3x3
PLC: QCA7500
WiFi: QCA9882 5G ac 2x2
Switch: QCA8337 3x1000M
Flash: 16MB 25L12835F SPI-NOR
DRAM SoC: 64MB w9751g6kb-25
DRAM PLC: 128MB w631gg6kb-15

Clocks: CPU:775.000MHz, DDR:650.000MHz, AHB:258.333MHz, Ref:25.000MHz

MAC addresses as verified by OEM firmware:
use address source
Lan/Wan/PLC *:10 art 0x1002 (label)
2G *:10 art 0x1000
5G *:14 art 0x5000

Important notes:

the PLC firmware has to be provided and copied manually onto the
device! The PLC here has no dedicated flash, thus the firmware file
has to be uploaded to the PLC controller at every system start
the PLC functionality is managed by the script /etc/init.d/plc_basic,
a very basic script based on the the one from Netadair (netadair dot de)
Installation:

Asus windows recovery tool:

have to have the latest Asus firmware flashed before continuing!
install the Asus firmware restoration utility
unplug the router, hold the reset button while powering it on
release when the power LED flashes slowly
specify a static IP on your computer:
IP address: 192.168.1.75
Subnet mask 255.255.255.0
start the Asus firmware restoration utility, specify the factory image
and press upload
do NOT power off the device after OpenWrt has booted until the LED flashing
TFTP Recovery method:

have to have the latest Asus firmware flashed before continuing!
set computer to a static ip, 192.168.1.75
connect computer to the LAN 1 port of the router
hold the reset button while powering on the router for a few seconds
send firmware image using a tftp client; i.e from linux:
$ tftp
tftp> binary
tftp> connect 192.168.1.1
tftp> put factory.bin
tftp> quit
do NOT power off the device after OpenWrt has booted until the LED flashing
Additional notes:

the pairing buttons have to have pressed for at least half a second,
it doesn't matter on which plc device (master or slave) first
it is possible to pair the devices without the button-pairing requirement
simply by pressing reset on the slave device. This will default to the
firmware settings, which is also how the plc_basic script is setting up
the master device, i.e. configuring it to firmware defaults
the PL-E56P slave PLC has its dedicated 4MByte SPI, thus it is capable
to store all firmware currently available. Note that some other
slave devices are not guarantied to have the capacity for the newer
~1MByte firmware blobs!
To have a good overlook about the slave device, here are its specs:
same QCA7500 PLC controller, same w631gg6kb-15 128MB RAM,
25L3233F 4MB SPI-NOR and an AR8035-A 1000M-Transceiver

Signed-off-by: Tamas Balogh <tamasbalogh@hotmail.com>

hauke commented:
You could add the script to extract and install the PLC firmware into an OpenWrt package like it is done for the lantiq vectoring VDSL firmware, see https://github.com/openwrt/openwrt/tree/master/package/kernel/lantiq/ltq-vdsl-fw/

hauke commented:
Thank you for your patch, I applied it to master in e1dcaeb55c939365cdd27e469b63fca295bf1a2e.
